File: apps/patter-voice-sdk/libraries/python/getpatter/services/fallback_provider.py
"""Fallback LLM provider that tries multiple providers in sequence.

If the primary provider fails, the next provider is tried, and so on.
Each provider gets a configurable number of retries before being skipped.
Failed providers are marked unavailable and periodically re-checked in the
background.
"""

from __future__ import annotations

__all__ = ["FallbackLLMProvider", "AllProvidersFailedError", "PartialStreamError"]

import asyncio
import logging
from typing import AsyncIterator

from getpatter.exceptions import PatterError
from getpatter.services.llm_loop import (
    LLMProvider,
    _stream_accepted_context_kwargs,
)

logger = logging.getLogger("getpatter")


# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class AllProvidersFailedError(PatterError):
    """Raised when every provider in the fallback chain has failed."""


class PartialStreamError(PatterError):
    """Raised when a provider fails after already yielding partial output."""


# ---------------------------------------------------------------------------
# FallbackLLMProvider
# ---------------------------------------------------------------------------


class FallbackLLMProvider:
    """LLM provider that tries multiple providers in sequence with failover.

    Args:
        providers: Ordered list of :class:`LLMProvider` instances.
        max_retry_per_provider: Number of attempts per provider before moving
            to the next (default ``1``).
        recovery_interval_s: Seconds between background recovery probes
            (default ``30``).
    """

    def __init__(
        self,
        providers: list[LLMProvider],
        *,
        max_retry_per_provider: int = 1,
        recovery_interval_s: float = 30.0,
    ) -> None:
        if not providers:
            raise ValueError("FallbackLLMProvider requires at least one provider")

        self._providers: list[LLMProvider] = list(providers)
        self._availability: list[bool] = [True] * len(providers)
        self._max_retry_per_provider = max_retry_per_provider
        self._recovery_interval_s = recovery_interval_s
        self._recovery_tasks: list[asyncio.Task[None] | None] = [None] * len(providers)

    # ------------------------------------------------------------------
    # Public helpers
    # ------------------------------------------------------------------

    def get_availability(self) -> list[bool]:
        """Return a snapshot of per-provider availability."""
        return list(self._availability)

    def destroy(self) -> None:
        """Cancel all background recovery tasks. Call on shutdown.

        Prefer :meth:`aclose` in async contexts — it awaits task cancellation
        and guarantees no pending tasks survive the owning event loop.
        """
        for i, task in enumerate(self._recovery_tasks):
            if task is not None:
                task.cancel()
                self._recovery_tasks[i] = None

    async def aclose(self) -> None:
        """Cancel probe tasks and await them. Safe to call multiple times."""
        tasks = [t for t in self._recovery_tasks if t is not None]
        for t in tasks:
            t.cancel()
        for t in tasks:
            try:
                await t
            except (asyncio.CancelledError, Exception):
                pass
        self._recovery_tasks = [None] * len(self._providers)

    async def __aenter__(self) -> "FallbackLLMProvider":
        return self

    async def __aexit__(self, *_exc) -> None:
        await self.aclose()

    # ------------------------------------------------------------------
    # LLMProvider implementation
    # ------------------------------------------------------------------

    async def complete_stream(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
        *,
        cancel_event: asyncio.Event | None = None,
        call_id: str | None = None,
    ) -> AsyncIterator[str]:
        """Stream only the text deltas, flattening the chunk envelope.

        Convenience wrapper over :meth:`stream` for callers that only want the
        assistant's text output and don't need tool-call or done markers.
        Mirrors the TypeScript SDK's ``fallback.completeStream`` shape.
        """
        async for chunk in self.stream(
            messages,
            tools,
            cancel_event=cancel_event,
            call_id=call_id,
            caller=caller,
            callee=callee,
        ):
            if chunk.get("type") == "text":
                yield chunk.get("content", "")

    async def stream(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
        *,
        cancel_event: asyncio.Event | None = None,
        call_id: str | None = None,
        caller: str | None = None,
        callee: str | None = None,
    ) -> AsyncIterator[dict]:
        """Try providers in sequence, yielding chunks from the first that succeeds.

        ``cancel_event`` (optional, set on barge-in by the stream handler) is
        forwarded unconditionally to each delegate provider's ``stream`` so a
        ``FallbackLLMProvider`` can be used as the pipeline provider — the
        built-in LLM loop calls ``provider.stream(..., cancel_event=...)`` on
        every turn, and a fallback that dropped the kwarg would raise
        ``TypeError`` on the first turn.

        ``call_id`` (optional, per-call session id) is forwarded the same way so
        session-aware delegate providers (Hermes / OpenClaw / OpenAI-compatible)
        still receive it when wrapped by the fallback.
        """
        errors: list[Exception] = []

        # First pass: try available providers
        async for chunk in self._try_providers(
            messages,
            tools,
            available_only=True,
            errors=errors,
            cancel_event=cancel_event,
            call_id=call_id,
            caller=caller,
            callee=callee,
        ):
            if isinstance(chunk, _Done):
                return
            yield chunk

        # All-failed fallback: retry every provider once more
        logger.warning(
            "FallbackLLMProvider: all providers unavailable, retrying all once"
        )
        async for chunk in self._try_providers(
            messages,
            tools,
            available_only=False,
            errors=errors,
            cancel_event=cancel_event,
            call_id=call_id,
            caller=caller,
            callee=callee,
        ):
            if isinstance(chunk, _Done):
                return
            yield chunk

        raise AllProvidersFailedError(
            f"All {len(self._providers)} LLM providers failed. "
            f"Last error: {errors[-1] if errors else 'unknown'}"
        )

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    async def _try_providers(
        self,
        messages: list[dict],
        tools: list[dict] | None,
        *,
        available_only: bool,
        errors: list[Exception],
        cancel_event: asyncio.Event | None = None,
        call_id: str | None = None,
        caller: str | None = None,
        callee: str | None = None,
    ) -> AsyncIterator[dict | _Done]:
        """Try each provider, yielding chunks or a _Done sentinel."""
        for i, provider in enumerate(self._providers):
            if available_only and not self._availability[i]:
                continue

            for attempt in range(self._max_retry_per_provider):
                try:
                    retry_label = f" (retry {attempt})" if attempt > 0 else ""
                    logger.info(
                        "FallbackLLMProvider: trying provider %d%s",
                        i,
                        retry_label,
                    )

                    yielded_tokens = False
                    # Only thread the per-call context kwargs the delegate's
                    # ``stream`` actually declares (or absorbs via **kwargs):
                    # a minimal custom provider without ``call_id`` would
                    # otherwise raise TypeError on every attempt and flap
                    # between unavailable and recovered forever. Session-aware
                    # delegates (Hermes caller_hash et al.) need caller/callee.
                    accepted = _stream_accepted_context_kwargs(provider)
                    ctx_kwargs = {
                        k: v
                        for k, v in (
                            ("call_id", call_id),
                            ("caller", caller),
                            ("callee", callee),
                        )
                        if k in accepted
                    }
                    async for chunk in provider.stream(
                        messages, tools, cancel_event=cancel_event, **ctx_kwargs
                    ):
                        yield chunk
                        yielded_tokens = True

                    # Success — restore availability if needed
                    if not self._availability[i]:
                        self._availability[i] = True
                        self._stop_recovery(i)
                        logger.info("FallbackLLMProvider: provider %d recovered", i)

                    yield _Done()
                    return

                except PartialStreamError:
                    raise

                except Exception as exc:
                    if yielded_tokens:
                        msg = (
                            f"FallbackLLMProvider: provider {i} failed after "
                            "yielding tokens — cannot retry"
                        )
                        logger.warning(msg)
                        raise PartialStreamError(msg) from exc

                    errors.append(exc)
                    logger.warning(
                        "FallbackLLMProvider: provider %d attempt %d failed — %s",
                        i,
                        attempt + 1,
                        exc,
                    )

            # Exhausted retries for this provider
            self._mark_unavailable(i)

    def _mark_unavailable(self, index: int) -> None:
        if not self._availability[index]:
            return
        self._availability[index] = False
        logger.warning("FallbackLLMProvider: marking provider %d as unavailable", index)
        self._start_recovery(index)

    def _start_recovery(self, index: int) -> None:
        if self._recovery_tasks[index] is not None:
            return

        async def _probe() -> None:
            while True:
                await asyncio.sleep(self._recovery_interval_s)
                try:
                    logger.debug(
                        "FallbackLLMProvider: probing provider %d for recovery",
                        index,
                    )
                    gen = self._providers[index].stream(
                        [{"role": "user", "content": "ping"}], None
                    )
                    # Drain one chunk to verify the provider responds.
                    # Explicit aclose() ensures the underlying HTTP stream
                    # (e.g. OpenAI SDK) is released even when the async
                    # generator early-exits before normal completion.
                    try:
                        async for _ in gen:
                            break
                    finally:
                        await gen.aclose()

                    self._availability[index] = True
                    self._stop_recovery(index)
                    logger.info("FallbackLLMProvider: provider %d recovered", index)
                    return
                except Exception:
                    pass  # Still unavailable — keep probing

        self._recovery_tasks[index] = asyncio.create_task(_probe())

    def _stop_recovery(self, index: int) -> None:
        task = self._recovery_tasks[index]
        if task is not None:
            task.cancel()
            self._recovery_tasks[index] = None


# ---------------------------------------------------------------------------
# Internal sentinel
# ---------------------------------------------------------------------------


class _Done:
    """Internal sentinel to signal successful completion."""

File: apps/patter-voice-sdk/libraries/typescript/src/fallback-provider.ts
/**
 * Fallback LLM provider that tries multiple providers in sequence.
 *
 * If the primary provider fails, the next provider is tried, and so on.
 * Each provider gets a configurable number of retries before being skipped.
 * Failed providers are marked unavailable and periodically re-checked in the
 * background.
 */

import type { LLMProvider, LLMChunk, LLMStreamOptions } from './llm-loop';
import { getLogger } from './logger';

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

/** Constructor options for `FallbackLLMProvider`. */
export interface FallbackLLMProviderOptions {
  /** Number of retry attempts per provider before moving to the next (default 1). */
  readonly maxRetryPerProvider?: number;
  /** Interval in ms between background recovery probes (default 30_000). */
  readonly recoveryIntervalMs?: number;
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

/** Thrown when all providers have been exhausted. */
export class AllProvidersFailedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'AllProvidersFailedError';
  }
}

/** Thrown when a provider fails after already yielding partial output. */
export class PartialStreamError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PartialStreamError';
  }
}

// ---------------------------------------------------------------------------
// FallbackLLMProvider
// ---------------------------------------------------------------------------

/** LLM provider that delegates to a sequence of underlying providers, falling back on failure. */
export class FallbackLLMProvider implements LLMProvider {
  private readonly providers: ReadonlyArray<LLMProvider>;
  private readonly availability: boolean[];
  private readonly maxRetryPerProvider: number;
  private readonly recoveryIntervalMs: number;
  private readonly recoveryTimers: Array<ReturnType<typeof setInterval> | null>;

  constructor(
    providers: ReadonlyArray<LLMProvider>,
    options?: FallbackLLMProviderOptions,
  ) {
    if (providers.length === 0) {
      throw new Error('FallbackLLMProvider requires at least one provider');
    }

    this.providers = providers;
    this.availability = providers.map(() => true);
    this.maxRetryPerProvider = options?.maxRetryPerProvider ?? 1;
    this.recoveryIntervalMs = options?.recoveryIntervalMs ?? 30_000;
    this.recoveryTimers = providers.map(() => null);
  }

  // -----------------------------------------------------------------------
  // Public helpers
  // -----------------------------------------------------------------------

  /** Returns a snapshot of per-provider availability. */
  getAvailability(): ReadonlyArray<boolean> {
    return [...this.availability];
  }

  /** Clears all background recovery timers. Call this when shutting down. */
  destroy(): void {
    for (let i = 0; i < this.recoveryTimers.length; i++) {
      if (this.recoveryTimers[i] !== null) {
        clearInterval(this.recoveryTimers[i]!);
        this.recoveryTimers[i] = null;
      }
    }
  }

  /**
   * Async-friendly disposer. Parity with Python's ``FallbackLLMProvider.aclose()``
   * — safe to call multiple times, returns a resolved Promise once all probe
   * timers are cleared. Prefer this in async contexts so awaiting the
   * shutdown integrates naturally with the owning lifecycle.
   */
  async aclose(): Promise<void> {
    this.destroy();
  }

  /**
   * Explicit-resource-management hook so callers can write
   * ``await using fallback = new FallbackLLMProvider([...])`` and have
   * background probe timers cleared automatically when the block exits.
   * Mirrors Python's ``async with FallbackLLMProvider(...)``.
   */
  async [Symbol.asyncDispose](): Promise<void> {
    await this.aclose();
  }

  /**
   * Stream only the text deltas, flattening the chunk envelope. Parity with
   * Python's ``FallbackLLMProvider.complete_stream``. Tool-call and done
   * markers are filtered out so callers can concatenate the yielded strings
   * directly.
   */
  async *completeStream(
    messages: Array<Record<string, unknown>>,
    tools?: Array<Record<string, unknown>> | null,
    opts?: LLMStreamOptions,
  ): AsyncGenerator<string, void, unknown> {
    for await (const chunk of this.stream(messages, tools, opts)) {
      if ((chunk as { type?: string }).type === 'text') {
        yield ((chunk as { content?: string }).content ?? '') as string;
      }
    }
  }

  // -----------------------------------------------------------------------
  // LLMProvider implementation
  // -----------------------------------------------------------------------

  /** Streaming entry point — yields chunks from the first provider that succeeds. */
  async *stream(
    messages: Array<Record<string, unknown>>,
    tools?: Array<Record<string, unknown>> | null,
    opts?: LLMStreamOptions,
  ): AsyncGenerator<LLMChunk, void, unknown> {
    const errors: Error[] = [];

    // First pass: try available providers
    const result = yield* this.tryProviders(
      messages,
      tools,
      /* availableOnly */ true,
      errors,
      opts,
    );
    if (result === 'done') return;

    // All-failed fallback: retry every provider once more
    getLogger().warn(
      'FallbackLLMProvider: all providers unavailable, retrying all once',
    );
    const retryResult = yield* this.tryProviders(
      messages,
      tools,
      /* availableOnly */ false,
      errors,
      opts,
    );
    if (retryResult === 'done') return;

    throw new AllProvidersFailedError(
      `All ${this.providers.length} LLM providers failed. Last error: ${errors.at(-1)?.message ?? 'unknown'}`,
    );
  }

  // -----------------------------------------------------------------------
  // Internals
  // -----------------------------------------------------------------------

  private async *tryProviders(
    messages: Array<Record<string, unknown>>,
    tools: Array<Record<string, unknown>> | null | undefined,
    availableOnly: boolean,
    errors: Error[],
    opts?: LLMStreamOptions,
  ): AsyncGenerator<LLMChunk, 'done' | 'exhausted', unknown> {
    for (let i = 0; i < this.providers.length; i++) {
      if (availableOnly && !this.availability[i]) continue;

      for (let attempt = 0; attempt < this.maxRetryPerProvider; attempt++) {
        try {
          getLogger().info(
            `FallbackLLMProvider: trying provider ${i}${attempt > 0 ? ` (retry ${attempt})` : ''}`,
          );

          let yieldedTokens = false;
          const gen = this.providers[i].stream(messages, tools, opts);

          // Consume and re-yield every chunk from the provider
          while (true) {
            let iterResult: IteratorResult<LLMChunk, void>;
            try {
              iterResult = await gen.next();
            } catch (err) {
              // Error during iteration — check if partial output was consumed
              if (yieldedTokens) {
                const msg = `FallbackLLMProvider: provider ${i} failed after yielding tokens — cannot retry`;
                getLogger().warn(msg);
                throw new PartialStreamError(msg);
              }
              throw err;
            }

            if (iterResult.done) break;

            yield iterResult.value;
            yieldedTokens = true;
          }

          // If we reach here, the provider succeeded — restore availability
          if (!this.availability[i]) {
            this.availability[i] = true;
            this.stopRecovery(i);
            getLogger().info(
              `FallbackLLMProvider: provider ${i} recovered`,
            );
          }

          return 'done';
        } catch (err) {
          if (err instanceof PartialStreamError) {
            throw err; // Do not retry
          }

          const error =
            err instanceof Error ? err : new Error(String(err));
          errors.push(error);
          getLogger().warn(
            `FallbackLLMProvider: provider ${i} attempt ${attempt + 1} failed — ${error.message}`,
          );
        }
      }

      // Mark provider as unavailable and start background recovery
      this.markUnavailable(i);
    }

    return 'exhausted';
  }

  private markUnavailable(index: number): void {
    if (!this.availability[index]) return; // already marked
    this.availability[index] = false;
    getLogger().warn(
      `FallbackLLMProvider: marking provider ${index} as unavailable`,
    );
    this.startRecovery(index);
  }

  private startRecovery(index: number): void {
    if (this.recoveryTimers[index] !== null) return; // already running

    this.recoveryTimers[index] = setInterval(async () => {
      try {
        getLogger().debug(
          `FallbackLLMProvider: probing provider ${index} for recovery`,
        );
        const gen = this.providers[index].stream(
          [{ role: 'user', content: 'ping' }],
          null,
        );
        // Drain one chunk to verify the provider responds, then close the generator
        try {
          await gen.next();
        } finally {
          await gen.return(undefined);
        }

        this.availability[index] = true;
        this.stopRecovery(index);
        getLogger().info(
          `FallbackLLMProvider: provider ${index} recovered`,
        );
      } catch {
        // Still unavailable — keep probing
      }
    }, this.recoveryIntervalMs);
  }

  private stopRecovery(index: number): void {
    if (this.recoveryTimers[index] !== null) {
      clearInterval(this.recoveryTimers[index]!);
      this.recoveryTimers[index] = null;
    }
  }
}
